src/tutorials/0008-merkle-dags/07.md

```diff
@@ -0,0 +1,53 @@
+---
+    title: "Merkle DAGs: Deduplication"
+    type: "multiple-choice" # can be "text", "code", "file-upload" or "multiple-choice"
+---
+
+Merkle DAGs offer a straightforward way of achieving *data
+deduplication*, efficiently storing data by encoding redundant
+sections as links. This can take place on both small and large
+scales.
+
+For an example of small-scale data duplication, consider the use
+case of tracking changes files in a directory over time (this is
+often called *versioning*). One change we could make to this
+directory is to delete the "fish" directory, replacing it with a
+directory called "dogs". These changes result in a new DAG,
+representing an updated state of the directory. However, all of
+the nodes representing the "cats" directory and its files are
+common to both DAGs. Therefore, we can reuse them, as depicted
+below, where the orange nodes represent nodes that are only used
+in the original DAG, the green nodes represent those that are
+common to both, and the blue nodes represent the extra nodes
+needed for the new state.
+
+This means that we can actually store both versions of the “pics”
+directory, without taking up twice as much space as it takes to
+store a single version! Git, a common version control system,
+uses Merkle DAGs in a very similar way to track changes to source
+code in software projects!
+
+Deduplication makes an even bigger difference when you expand
```

Comment:
From the Google Doc:

Does "pull the new nodes" mean end users of the website downloading just bits of new code instead of all of it in order to view the website? I don't think "normal people" believe they're pulling or downloading anything when they view a website, so this one's a little harder for me to wrap my head around. Is there an example where someone would more knowingly be avoiding downloading/adding unnecessary stuff?

Reply:
<this comment is part rubber-ducking for myself>
Two issues here: the first is that users don't understand that their web browser downloads content. Maybe we could avoid that by making it explicit, or else think of something that people explicitly download by clicking to do so. For the latter, the challenge is finding something that people explicitly download many instances of, with minor variations. Something that immediately comes to mind is downloading the same word document shared over email, but...

...the second issue is that the original intent behind this example (I believe Rodd suggested it) was to get into talking about scale: e.g., Merkle forest style. Massive, Internet-scale cross-pollination, millions of sites using variations on the same theme. Different DAGs all borrowing from one another.

... the third issue is that the example right now is inter-directory deduplication, rather than inter-file deduplication, which isn't explained until the last paragraph. I'm having trouble coming up with good example of inter-directory deduplication that don't seem esoteric to your average internet user.

So... putting it all together, what about memes as an example? If you download every variant of a meme you come across, the picture is the same but the text is different. (this is kind of a lie because that's not how all image formats work, but maybe that's okay). To make this work, we need to move the last paragraph up.

Reply:
@mitchwagner Regarding the CSS example, would it lose its impact for you if "pull" were changed to "access"? eg "As users visit websites that depend on variants of this theme, they'd only need to access the new nodes, leading to extra speedy load times!" This is a way I could envision to get across the benefit to the end user without them thoroughly understanding how the process works. 

I'm fine with having multiple examples that make sense to people at different levels of understanding. 

The meme example could potentially work, and I agree you'd want to reorder if you did it. 

Perhaps @ribasushi has suggestions?

Reply:
This might seem like splitting hairs, but to me, "pull" connotates local storage/caching where access doesn't, and saying a user only needs to access the new nodes is wrong: they need to access the old nodes and the new nodes.

What if we're just more explicit? Proposal:

Deduplication makes an even bigger difference when you expand this concept to a larger scale. For example, consider the use case of browsing the web! When a person visits a web page using a browser, the browser must first download any resources associated with the page, including images, text, and styling markup. You may have noticed that a lot of web pages actually look fairly similar, using slightly different variants of a common theme. Often, there's a lot of redundancy here - the themes are mostly the same, with tweaks here and there in the data that describes them.

On the location-addressed web, each of these themes, despite sharing lots of the same data, is often completely re-downloaded, as there's no easy way to identify this redundancy. In contrast, if Merkle DAGs were used to distribute the themes, they'd all share an identifiable common core that is unchanged from one theme to the next, and so the browser could be smart enough to avoid downloading this component more than once. Whenever a user visited a new website using a variation of the theme, the browser would only need to download the nodes of its DAG corresponding to the parts that are different, having already previously downloaded the rest! For resources that are used extremely widely on the Internet (think WordPress themes, the Bootstrap CSS library, or common JavaScript libraries), this could result in tremendous reduction in wasteful, needless downloads!

Reply:
@mitchwagner I like it!

Reply:
Will commit, then!